src/Agent/NewRelic/Agent/Core/Agent.cs

```diff
@@ -562,19 +567,82 @@ internal void Detach(bool removeAsync, bool removePrimary)
 
             var filteredContextData = new Dictionary<string, object>();
 
-            // first, handle the "includes"
-            if (_configurationService.Configuration.ContextDataInclude.Count() == 0) // empty "include" list, so include everything
+            // Pseudocode
+            // Include/exclude rules:
+            // 1. Exclude wins over include
+            // 2. Can have a '*' wildcard at the end of an include/exclude definition
+            // 3. "more specific" rules win over "less specific".  Example: include=AB, exclude=A*, AB would be included but not ABC or AAA
+
+            // Algorithm:
+            // Create list of rules, sorted by least specific to most specific, with includes first and then excludes within each level of specificity
+            // Specificity is just the length of the string, minus any wildcard character at the end
+            // Apply rules in order. Include rules add attributes to filteredContextData from unfilteredContextData, and exclude rules remove attributes from filteredContextData
+
+            var allRulesInOrder = _orderedCludeRuleList ?? (_orderedCludeRuleList = GetOrderedCludeRuleList()); 
+
+            if (_includeRuleList.Count() == 0) // empty "include" list, so include everything
             {
                 filteredContextData = unfilteredContextData;
             }
-            else
+
+            // Now apply the rules in order
+            // TODO: maybe optimize by just working with keys as lists of strings until it comes time to return the actual data
+            foreach (var rule in allRulesInOrder)
             {
-                filteredContextData = unfilteredContextData.Where(property => _configurationService.Configuration.ContextDataInclude.Contains(property.Key)).ToDictionary();
+                if (rule.Include)
+                {
+                    filteredContextData = filteredContextData.Concat(unfilteredContextData.Where(x => Regex.IsMatch(x.Key, rule.Text))).ToDictionary();
```

Comment:
I'm not sure about using `Regex.IsMatch` for a few reasons:
* These are simple string comparisons (either `Equals` or `StartsWith`) and regex will be slower
* It will break on special characters
* It allows users to actually use a regex, which isn't the intent

Reply:
Thanks for pointing this out; I've fixed it.